File: tests/FSharpx.Tests/PartialLensTests.fs
module FSharpx.Tests.PartialLensTests

open FSharpx
open NUnit.Framework

type StateCity = {
    State: string
    City: string option
} with 
    static member cityP =
        let get (a: StateCity) = a.City
        let set (a: StateCity) city = { a with City = Some city }
        PartialLens.fromGetSet get set

type Search = {
    Name: string option
    StateCity: StateCity option
} with
    static member stateCityP =
        let get (a: Search) = a.StateCity
        let set (a: Search) sc = { a with StateCity = Some sc }
        PartialLens.fromGetSet get set

let searchCity = Search.stateCityP |> PartialLens.compose StateCity.cityP

[<Test>]
let setAllSome() =
    let allSome = 
        { Name = Some "John"; StateCity = Some { State = "Florida"; City = Some "Miami" } }
    let orlando = searchCity |> PartialLens.set "Orlando" allSome
    Assert.AreEqual(Some "Orlando", orlando.StateCity.Value.City)
    
[<Test>]
let setNoCity() =
    let noCity = 
        { Name = Some "John"; StateCity = Some { State = "Florida"; City = None } }
    let orlando = searchCity |> PartialLens.set "Orlando" noCity
    Assert.AreEqual(None, orlando.StateCity.Value.City)

[<Test>]
let setNoState() =
    let noState = 
        { Name = Some "John"; StateCity = None }
    let orlando = searchCity |> PartialLens.set "Orlando" noState
    Assert.AreEqual(None, orlando.StateCity)

File: tests/FSharpx.CSharpTests/PartialLensTests.cs
using System;
using Microsoft.FSharp.Core;
using NUnit.Framework;
using FSharpx;

namespace FSharpx.CSharpTests {
    [TestFixture]
    public class PartialLensTests {
        private class StateCity {
            public readonly string State;
            public readonly FSharpOption<string> City;

            public static readonly PartialLens<StateCity, string> CityP =
                PartialLensEx.Create((StateCity x) => x.City, (x, v) => new StateCity(x.State, v.Some()));

            public StateCity(string state, FSharpOption<string> city) {
                State = state;
                City = city;
            }
        }

        private class Search {
            public readonly FSharpOption<string> Name;
            public readonly FSharpOption<StateCity> StateCity;

            public static readonly PartialLens<Search, StateCity> StateCityP =
                PartialLensEx.Create((Search x) => x.StateCity, (x,v) => new Search(x.Name, v.Some()));

            public static readonly PartialLens<Search, string> CityP = StateCityP.AndThen(PartialLensTests.StateCity.CityP);

            public readonly InstancePartialLens<Search, string> ICityP;

            public Search(FSharpOption<string> name, FSharpOption<StateCity> stateCity) {
                Name = name;
                StateCity = stateCity;
                ICityP = InstancePartialLens.Create(this, CityP);
            }
        }

        [Test]
        public void SetNoState() {
            var noState = new Search("John".Some(), null);
            var orlando = noState.ICityP.Set("Orlando");
            Assert.IsNull(orlando.StateCity);
        }
    }
}
